syntax = "proto3";

message Int {
  int64 data = 1;
}

message Long {
  int64 data = 1;
}

message String {
  string data = 1;
}

message Double {
  double data = 1;
}

// Empty used to represent a None object
message Empty {
}

message Bool {
  bool data = 1;
}

message Ref {
  uint64 data = 1;
}

message PyObj {
  bytes data = 1;
}

// Used for shipping remote functions to workers
message Function {
  bytes implementation = 1;
}

message ReusableVar {
  string name = 1; // The name of the reusable variable.
  Function initializer = 2; // A serialized version of the function that initializes the reusable variable.
  Function reinitializer = 3; // A serialized version of the function that reinitializes the reusable variable.
}

// Union of possible object types
message Obj {
  String string_data = 1;
  Int int_data = 2;
  Long long_data = 12;
  Double double_data = 3;
  Bool bool_data = 10;
  Tuple tuple_data = 7;
  List list_data = 4;
  Dict dict_data = 8;
  Array array_data = 5;
  Empty empty_data = 9;
  Ref objref_data = 11;
  PyObj pyobj_data = 6;
}

message List {
  repeated Obj elem = 1;
}

message Tuple {
  repeated Obj elem = 1;
}

message DictEntry {
  Obj key = 1;
  Obj value = 2;
}

message Dict {
  repeated DictEntry elem = 1;
}

message Value {
  uint64 ref = 1; // For pass by reference
  Obj obj = 2; // For pass by value
}

message Array {
  repeated uint64 shape = 1;
  sint64 dtype = 2;
  bool is_scalar = 8;
  repeated double double_data = 3;
  repeated float float_data = 4;
  repeated sint64 int_data = 5;
  repeated uint64 uint_data = 6;
  repeated uint64 objref_data = 7;
}
